### PR TITLE
Removes Box from Typography package

### PR DESCRIPTION
**Pull request by brookescarlett**
## ✍️ Proposed changes

Removes Box from Typography elements in favor of 'Polymorphic'
Adds Polymorphic to Label, to allow users to set the as prop

🎟 _Jira ticket:_ [LG-1678](https://jira.mongodb.org/browse/LG-1678)
🎟 _Jira ticket:_ [LG-2727](https://jira.mongodb.org/browse/LG-2727)

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Style change (purely visual updates; if this is the case, attach a screenshot below!)

## ✅ Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have run `yarn changeset` and documented my changes

- [ ] I have added my new package to the global tsconfig
- [ ] I have added my new package to the Table of Contents on the global README

## Feedback Levels

Please try to mark your comments with one of these levels

1. Praise (👏 🚀 🔥 🍾)
2. Opinions, Ideas, Suggestions (💡 👀)
3. Let’s discuss @ Code review (🤷 🤔 🧐 )
4. Doesn’t meet agreed upon standards (😬 ⚠️ ‼️)
5. Missing an edge case (🛑 🙅 ❌)
6. Missing core functionality (⛔️ ☣️ 🚧 🌋 💀)

(Note: Use ⌘-⌃-␣ to open the Emoji keyboard!)

### CODE REVIEW

**Comment by TheSonOfThomp** [comment]
@brookescarlett Can you merge in main? There are some conflicts with `Polymorphic`

File: packages/typography/src/Link/Link.tsx [hunk 1]
```diff
@@ -21,84 +25,89 @@ import {
...
+        />
+      );
+    }
 
-  const elementProps = href
-    ? ({ as: 'a', href, target, rel } as const)
-    : ({ as: 'span' } as const);
+    const elementProps = href
+      ? ({ as: 'a', href, target, rel } as const)
+      : ({ as: 'span' } as const);
```
Comment (TheSonOfThomp) [inline packages/typography/src/Link/Link.tsx:90]: We shouldn't need to pass in `as` to the `Component` element

Comment (TheSonOfThomp) [reply packages/typography/src/Link/Link.tsx:90]: Should move this logic above the, and call `usePolymorphic(href ? 'as' : 'span')`

File: packages/typography/src/Overline/Overline.tsx [hunk 2]
```diff
@@ -1,32 +1,41 @@
...
-        baseTypographyStyles,
-        overlineStyles,
-        defaultTextColor[theme],
-        className,
-      )}
-      {...rest}
-    />
-  );
-};
+export const Overline = Polymorphic<OverlineProps>(
```
Comment (TheSonOfThomp) [inline packages/typography/src/Overline/Overline.tsx:16]: Should `Disclaimer` also extend Polymorphic?

File: packages/typography/src/Label/Label.tsx [hunk 3]
```diff
@@ -13,32 +18,36 @@ import {
...
-  baseFontSize: baseFontSizeOverride,
-  darkMode: darkModeProp,
-  className,
-  children,
-  disabled = false,
-  ...rest
-}: LabelProps) => {
-  const baseFontSize = useUpdatedBaseFontSize(baseFontSizeOverride);
-  const { theme } = useDarkMode(darkModeProp);
+export const Label = Polymorphic<LabelProps>(
```
Comment (TheSonOfThomp) [inline packages/typography/src/Label/Label.tsx:21]: Should `Description` also extend Polymorphic?

**Review by TheSonOfThomp — APPROVED** [review]
Once the `Link` test is passing , everything else looks good 👍

File: packages/typography/src/Overline/Overline.tsx (hunk 2 above)
Comment (brookescarlett) [reply packages/typography/src/Overline/Overline.tsx:16]: We had a ticket to convert Label which is why I did.. happy to add Disclaimer if you feel it's necessary?

File: packages/typography/src/Label/Label.tsx (hunk 3 above)
Comment (brookescarlett) [reply packages/typography/src/Label/Label.tsx:21]: see above

**Comment by brookescarlett** [comment]
Fixed the failing test... PTAL @TheSonOfThomp

File: packages/typography/src/Link/Link.tsx [hunk 4]
```diff
@@ -21,84 +21,87 @@ import {
...
+    ...rest
+  }) => {
+    const [currentHostname, setCurrentHostname] = useState('');
+    useEffect(() => {
+      setCurrentHostname(window.location.hostname);
+    }, []);
 
-  const { theme } = useDarkMode(darkModeProp);
+    const { theme } = useDarkMode(darkModeProp);
+    const { Component } = usePolymorphic(as ? as : href ? 'a' : 'span');
```
Comment (TheSonOfThomp) [inline packages/typography/src/Link/Link.tsx:43]: Nit: This is a little more readable, imo: `as ?? href ? 'a' : 'span'`

File: packages/polymorphic/src/Polymorphic/Polymorphic.tsx [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
```
Comment (TheSonOfThomp) [inline packages/polymorphic/src/Polymorphic/Polymorphic.tsx:1]: We can remove this file. It was renamed `Polymorphic.ts`

**Review by TheSonOfThomp — APPROVED** [review]
Let's remove `Polymorphic.tsx`. LGTM otherwise 👍

File: packages/typography/src/Overline/Overline.tsx (hunk 2 above)
Comment (TheSonOfThomp) [reply packages/typography/src/Overline/Overline.tsx:16]: Don't think it's necessary, no. Just wanted to make sure 👍

**Pull request merged**